[clang][CodeGen] Generate follow-up metadata for loops in correct format

kasuga-fj opened the pull request:
When pragma of loop transformations is specified, follow-up metadata for loops is generated after each transformation. On the LLVM side, follow-up metadata is expected to be a list of properties, such as the following:

```
!followup = !{!"llvm.loop.vectorize.followup_all", !mp, !isvectorized}
!mp = !{!"llvm.loop.mustprogress"}
!isvectorized = !{"llvm.loop.isvectorized"}
```

However, on the clang side, the generated metadata contains an MDNode that has those properties, as shown below:

```
!followup = !{!"llvm.loop.vectorize.followup_all", !loop_id}
!loop_id = distinct !{!loop_id, !mp, !isvectorized}
!mp = !{!"llvm.loop.mustprogress"}
!isvectorized = !{"llvm.loop.isvectorized"}
```
According to the [LangRef](https://llvm.org/docs/TransformMetadata.html#transformation-metadata-structure), the LLVM side is correct. Due to this inconsistency, follow-up metadata was not interpreted correctly, e.g., only one transformation is applied when multiple pragmas are used.

This patch fixes clang side to emit followup metadata in correct format.

kasuga-fj commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -317,6 +317,35 @@ std::optional<MDNode *> llvm::makeFollowupLoopID(
+      // instead of !3, traverse all of Option's operands and push them into
+      // MDs if Option seems to be a LoopID.
+      if (auto *MDN = dyn_cast<MDNode>(Option)) {
+        // TODO: Is there a proper way to detect LoopID?
+        if (MDN->getNumOperands() > 1 && MDN->getOperand(0) == MDN) {
```
I don't think this is a very good implementation, but I couldn't think of a better way.

It may be better to force the followup metadata to have exactly one LoopID. As far as I've investigated, some tests (e.g., `Transforms/LoopDistribute/followup.ll`) contain followup metadata with non-LoopID data, but I believe we can fix them easily. Also, if my understanding is correct, those which actually generated by the frontend seem to have exactly one LoopID. Would it make sense to set such a restriction?

Meinersbur commented:
Looks to me that the mistake is that the `followup_all` contains the LoopID, while it should contain the individual properties. See e.g. https://github.com/llvm/llvm-project/blob/29925b7044517bcd1b3f76333a598092427c3636/llvm/test/Transforms/LoopDistribute/followup.ll#L129-L132 and the documentation at https://llvm.org/docs/TransformMetadata.html.

Sorry, didn't notice that when reviewing the test cast. Whatever created it should be fixed. Whether an MDNode is a LoopID cannot be detected reliably. Having the first entry refer to itself was a technique to avoid collapsing identical LoopIDs before we had `distinct`. There might be other kinds of metadata that also still use this technique. For LoopID it is unnecessary anyway, passes that duplicate code (e.g. inlining, LoopUnroll, ...) do not care about ensuring that each LoopID is distinct anyway. It should be more thought of as a bag of properties.

Meinersbur commented:
Looks like the problem is here: https://github.com/llvm/llvm-project/blob/39ce99589b05ddd43e5c8b7f89ab5bbfda40d2ce/clang/lib/CodeGen/CGLoopInfo.cpp#L134

It should add the properties, not the generated LoopID (that would be used if the #pragma was not the followup of another #pragma). Considering that it might be necessary to merge multiple such followup properties (eg. both of `followup_all` and `followup_vectorized` should be added to the vectorized loop).

Considering that I wrote both sides, I obviously wasn't consistent myself. If you fix it, could you fix it for wall `followup_*`? Possibly the `create*Metadata` should only return a list of properties instead of an MDNode, with the MDNode only to be created when applying the properties to a Loop. However, the easier fix would probably be to extract the properties from the MDNode and append to the `followup_*` MDNode, with the returned MDNode becoming garbage-collected at some point.

kasuga-fj commented:
> If you fix it, could you fix it for wall followup_*?

Yes, of course, I was going to fix everything from the beginning (in that sense, I should have added more tests in #131337). Thanks also for the implementation advice.

kasuga-fj committed:
[clang][CodeGen] Generate follow-up metadata for loops in correct format

When pragma of loop transformations is specified, follow-up metadata for
loops is generated after each transformation. On the LLVM side,
follow-up metadata is expected to be a list of properties, such as the
following:

```
!followup = !{!"llvm.loop.vectorize.followup_all", !mp, !isvectorized}
!mp = !{!"llvm.loop.mustprogress"}
!isvectorized = !{"llvm.loop.isvectorized"}
```

However, on the clang side, the generated metadata contains an MDNode
that has those properties, as shown below:

```
!followup = !{!"llvm.loop.vectorize.followup_all", !loop_id}
!loop_id = distinct !{!loop_id, !mp, !isvectorized}
!mp = !{!"llvm.loop.mustprogress"}
!isvectorized = !{"llvm.loop.isvectorized"}
```

According to the LangRef, the LLVM side is correct. (ref:
https://llvm.org/docs/TransformMetadata.html#transformation-metadata-structure).
Due to this inconsistency, follow-up metadata was not interpreted
correctly, e.g., only one transformation is applied when multiple
pragmas are used.

This patch fixes clang side to emit followup metadata in correct format.

kasuga-fj commented on clang/test/CodeGenCXX/pragma-loop.cpp:
```diff
@@ -227,36 +226,26 @@ void for_test_scalable_1(int *List, int Length) {
-// CHECK: ![[LOOP_6]] = distinct !{![[LOOP_6]], [[MP]], ![[WIDTH_2:.*]], ![[FIXED_VEC]], ![[INTERLEAVE_2:.*]], ![[FOLLOWUP_VECTOR_6:.*]]}
-// CHECK: ![[FOLLOWUP_VECTOR_6]] = !{!"llvm.loop.vectorize.followup_all", ![[AFTER_VECTOR_6:.*]]}
-// CHECK: ![[AFTER_VECTOR_6]] = distinct !{![[AFTER_VECTOR_6]], ![[ISVECTORIZED:.*]], ![[UNROLL_8:.*]]}
+// CHECK: ![[LOOP_6]] = distinct !{![[LOOP_6]], [[MP]], ![[WIDTH_2:.*]], ![[FIXED_VEC]], ![[INTERLEAVE_2:.*]], ![[FOLLOWUP_VECTOR_3]]}
```
I don't fully understand why `[[MP]]` was not in `![[FOLLOWUP_VECTOR_6]]` before this patch. Is it intentional?

kasuga-fj commented on clang/lib/CodeGen/CGLoopInfo.cpp:
```diff
@@ -44,23 +44,19 @@ MDNode *LoopInfo::createPipeliningMetadata(const LoopAttributes &Attrs,
           MDNode::get(Ctx, {MDString::get(Ctx, "llvm.loop.pipeline.disable"),
                             ConstantAsMetadata::get(ConstantInt::get(
                                 llvm::Type::getInt1Ty(Ctx), 1))}));
-      LoopProperties = NewLoopProperties;
```
Different from this PR, but I think we should set `HasUserTransforms` to true here (same for other `create*Metadata`). If not, the user-specified `disable' attributes would not be generated properly.

kasuga-fj replied:
What about other transformations, e.g., vectorization?
https://github.com/llvm/llvm-project/blob/ce8febb0befe41694b9d83c14dcfb831a82489ff/clang/lib/CodeGen/CGLoopInfo.cpp#L213-L220

I just looked for it and found an issue that might be caused by this.
https://github.com/llvm/llvm-project/issues/75839

kasuga-fj commented on clang/lib/CodeGen/CGLoopInfo.cpp:
```diff
@@ -170,11 +161,10 @@ LoopInfo::createUnrollAndJamMetadata(const LoopAttributes &Attrs,
+      Attrs, FollowupLoopProperties, FollowupHasTransforms);
 
   SmallVector<Metadata *, 4> Args;
-  Args.push_back(nullptr);
```
`nullptr` is used as a placeholder for LoopID, so this is no longer necessary, right?

kasuga-fj reviewed (commented):
>  Possibly the `create*Metadata` should only return a list of properties instead of an MDNode, with the MDNode only to be created when applying the properties to a Loop. However, the easier fix would probably be to extract the properties from the MDNode and append to the `followup_*` MDNode, with the returned MDNode becoming garbage-collected at some point.

Fixed `create*Metadata` to return a list of properties instead of an MDNode. I believe unnecessary MDNodes would not be generated.

Meinersbur reviewed (approved):
LGTM, great work!

kasuga-fj commented on clang/lib/CodeGen/CGLoopInfo.cpp:
```diff
@@ -44,23 +44,19 @@ MDNode *LoopInfo::createPipeliningMetadata(const LoopAttributes &Attrs,
   if (Enabled != true) {
-    SmallVector<Metadata *, 4> NewLoopProperties;
     if (Enabled == false) {
-      NewLoopProperties.append(LoopProperties.begin(), LoopProperties.end());
```
When `Enabled == std::nullopt`, `LoopProperties` was used as is, not `NewProperties`. So I think the cause is elsewhere. Anyway, it's enough to know that `llvm.mustprogress` should be appended unconditionally, thanks.

Meinersbur replied:
In principle, LoopVectorize should know that if the original loop had a progress guarantee, then the vectorized loop will as well, so it should add `llvm.loop.must_progress` no matter what. I don't think it currently does.

kasuga-fj replied:
> In principle, LoopVectorize should know that if the original loop had a progress guarantee, then the vectorized loop will as well, so it should add `llvm.loop.must_progress` no matter what.

I think this is completely correct. What I didn't understand is, why the followup metadata of `LOOP_6` (`FOLLOW_VECTOR_6`) didn't have `llvm.mustprogress` before this patch, but now it (`FOLLOWUP_VECTOR_3`) does.  I investigated a little deeper and found the cause; `FOLLOWUP_VECTOR_6` actually had `mustprogress` (?!). That is, the test passed for both of the following directives.

```
// Original.
// CHECK: ![[AFTER_VECTOR_6]] = distinct !{![[AFTER_VECTOR_6]], ![[ISVECTORIZED:.*]], ![[UNROLL_8:.*]]}

// This was also fine.
// CHECK: ![[AFTER_VECTOR_6]] = distinct !{![[AFTER_VECTOR_6]], [[MP]], ![[ISVECTORIZED:.*]], ![[UNROLL_8:.*]]}
```

Maybe FileCheck has a problem?

kasuga-fj replied:
The same thing seems to happen elsewhere, e.g. `LOOP_6` actually has `vectorize.enable` but is not included in the CHECK directive.

Meinersbur replied:
I think this is due to the regex `.*` being too greedy, so e.g. `![[ISVECTORIZED:.*]]` consumes multiple metadata nodes. A better one would be `[_a-zA-Z0-9.]+`. In principle `![[ISVECTORIZED]] = {!"llvm.loop.isvectorized"}` (or `UNROLL_8`) should be verified somewhere, which would then fail if it matched more than one node.

IMHO there are lots of problems with FileCheck on LLVM-IR, and this is just one of them. Another one is that by default `CHECK: car sale` will match `carpet sale`.

kasuga-fj replied:
> I think this is due to the regex .* being too greedy, so e.g. ![[ISVECTORIZED:.*]] consumes multiple metadata nodes.

That makes sense! I got it, thank you!

> Another one is that by default `CHECK: pet store` will match `carpet store`.

Ugh, that's a tricky problem.

kasuga-fj reviewed (commented):
Thanks for your review!

mrkajetanp commented:
For more details on the above, the problem happens once the loop vectorizer transforms a loop with this metadata that's inside an openmp block. The dbg metadata becomes incorrectly attached and as a result verifyFunction returns a failure.
```
!dbg attachment points at wrong subprogram for function
!2094 = distinct !DISubprogram(name: "main.omp_outlined", scope: !2, file: !2, line: 207, type: !2095, scopeLine: 207, flags: DIFlagArtificial | DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagLocalToUnit | DISPFlagDefinition | DISPFlagOptimized, unit: !25, retainedNodes: !2102)
ptr @main.omp_outlined
  br i1 %53, label %middle.block, label %vector.body, !dbg !2327, !llvm.loop !2338
!2339 = !DILocation(line: 258, column: 5, scope: !2162)
!2162 = distinct !DILexicalBlock(scope: !2163, file: !2, line: 258, column: 5)
!2118 = distinct !DISubprogram(name: "main.omp_outlined_debug__", scope: !2, file: !2, line: 208, type: !2095, scopeLine: 208, flags: DIFlagArtificial | DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagLocalToUnit | DISPFlagDefinition | DISPFlagOptimized, unit: !25, retainedNodes: !2119)
clang-21: ~/llvm/llvm-project-release/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:11013: bool llvm::LoopVectorizePass::processLoop(Loop *): Assertion `!verifyFunction(*L->getHeader()->getParent(), &dbgs())' failed.
```

kasuga-fj commented:
Thanks for sharing. I haven't looked at the details yet, but I simplified it as follows.

```c
#include <omp.h>

void g(float);

void f(int m) {
#pragma omp parallel
  {
    float v0 = 0.0;
#pragma omp simd reduction(+:v0)
#pragma unroll(4)
    for(int i=0; i < m; i++)
    {
      v0 += i;
    }

#pragma omp barrier
    g(v0);
  }
}

```

godbold: https://godbolt.org/z/P3cW4To99

Meinersbur commented:
> ```c
> #pragma omp simd reduction(+:v0)
> #pragma unroll(4)
> ```

That's not supposed to work. I think in the past the unroll pragma was just dropped. If the debug info is the issue, there hasn't been a lot of care to get OpenMP emit correct debug info, see e.g. #110700.  

For this reason OpenMP 5.1 added new pragma:
```c
#pragma omp simd reduction(+:v0)
#pragma omp unroll partial(4)
```

kasuga-fj commented:
> For this reason OpenMP 5.1 added new pragma:
> 
> ```c
> #pragma omp simd reduction(+:v0)
> #pragma omp unroll partial(4)
> ```

FWIW, I tried it and the following error occurred.

```
<source>:10:1: error: OpenMP constructs may not be nested inside a simd region except for ordered simd, simd, scan, or atomic directive
   10 | #pragma omp unroll partial(4)
      | ^
```

pawosm-arm commented:
One thing that worries me is that before this commit, there was no ICE caused by compiling this snippet of code (regardless its validity or usefulness), and now there is. It would be great if it could print a normal error message (which would encourage my resistant colleagues to rewrite the code), rather than seeing segfault which triggers the 'fix the compiler' demands.

kasuga-fj commented:
Printing an error message looks reasonable to me, but I think we first need to clarify what codes are "unsupported".
